### ISSUE TITLE
Translate and test the whole set of libskycoin API

### ISSUE THREAD

**Issue by olemis** [skycoin/skycoin#1191]
- [x] Wrappers for `libskycoin` API functions generated with https://github.com/simelo/cgogen (ensure it compiles).
- [x] Translate skycoin test suite from go to c (see simelo/cgogen#1) and ensure all test cases will pass when running it against static and shared versions of `libskycoin` .
- [x] Test against cipher golden file(s) test suite , see #1436 
- [x] Changes in test suite due to changes in `catchApiPanic` (after #1649 )
- [x] `Ripemd160` hash methods in cipher crypto API (somehow missing ...)

**Comment by olemis** [comment]
A summary after closing #1221 as duplicate of this issue 

Some functions depend on types that need to be converted to C in order for the function to be converted

Functions that depend on external libraries:

```
    Visor/Visor/* depends on bolt.DB "github.com/boltdb/bolt"
    Daemon/Daemon/* depends on Visor
    api/cli/cli.NewApp depends on App, App depends on App "github.com/urfave/cli"
    gui/gui/CreateHTTPS depends on Daemon
    util/elapse/elapse/* depends on logging.Logger
    util/http/httphelper/* depends on logging.Logger
    util/logging/logging/* depends logging.Logger "github.com/sirupsen/logrus"
```

Function that depend on GO standard libraries

```
    cipher/secp256k1-go/secp256k1-go2/secp256k1go/* depend on Number, Number depends on big/Math
    daemon/gnet/gnet/* depends on bytes.Buffer, net.Conn, time.Time
    daemon/pex/pex/* depends on time.Duration
    gui/gui/CSRFCheck depend sync.RWMutex
    wallet/wallet/* depend on sync.sync.RWMutex
```

**Comment by olemis** [comment]
@stdevEclipse pls limit initially the scope of this issue to files in `src/api/cli` and `src/cipher` packages . Can you deliver those already ?

**Comment by stdevEclipse** [comment]
Issues with libc:
- Function cipher.pbkdf2.Key takes parameter of type (func() hash.Hash). How to deal with parameters of type function when exporting to c lib?
- Functions in cipher.encoder take parameter of type interface. How to deal with this when exporting to c lib.

**Comment by gz-c** [comment]
not necessary to translate those

**Comment by stdevEclipse** [comment]
OK. Excellent.

**Comment by stdevEclipse** [comment]
Libskycoin set of functions for cipher and cli are complete. Wrapper functions use Handles for specific types like webrpc_client, wallet, app, config, gcli_context.
It still needs automatic tests.

**Comment by stdevEclipse** [comment]
Greetings.
I am trying to run "make test-libc", but I get an error saying that include file criterion/criterion.h is not found.
I am testing on Windows 10 using Docker.
Help, please.

**Comment by gz-c** [comment]
@olemis @iketheadore I think we'd need a separate developer container for @stdevEclipse 's use case.  This one would not need to go into docker hub if it would complicate things

**Comment by olemis** [comment]
@gz-c  the container has been built already by @stdevMauricio as part of #1073 . This is work in progress and shall be finished once we finish with a number of open issues . It will be published as `skycoin/skycoin-dev` in [Docker Hub](https://hub.docker.com/u/skycoin) .

**Comment by olemis** [comment]
@stdevEclipse current version of that image already includes a working installation of `Criterion` configured to run tests . Please continue with other tasks until your environment gets setup with this .

**Comment by stdevEclipse** [comment]
@gz-c, libc integration tests are having the following issues:

1- Creating a wallet with a specific seed.
  Creating a wallet with a specific seed, deleting and unloading the wallet, and the creating another wallet with the same seed raises an error with message "There is already a wallet with that seed". This shouldn't happen. If some one creates a wallet with a given seed, and then deletes it, they should be able to create another wallet with the same seed.
  My guess is that the problem lies on  map field named "Service.firstAddrIDMap", it should be updated when a wallet is unloaded. It is stil having a reference to the deleted wallet seed.

2- Encrypting or decrypting wallets.
   When calling encrypt or decrypt wallet methods more than once the server stops responding. Reported errors are connection refused or EOF. I am not sure about this problem, maybe it is a issue because of running Docker on Windows. There are several tests that encrypt or decrypt wallets, all these tests pass if they are executed separately. They only fail when running more than one of these tests.

**Comment by olemis** [comment]
About #2 @stdevEclipse could you please uncomment those and run'em inside the `skycoin-dev` Docker container ?

**Comment by stdevEclipse** [comment]
Do you mean commiting and testing with travis? Yes, I can do that.

**Comment by olemis** [comment]
Test inside the Docker `skycoin/skycoin-dev` Docker container . There's a GNU/Linux OS inside .

**Comment by stdevEclipse** [comment]
I am testing inside a containit with some version of Linux using Docker for Windows. I should try in docker for Ubuntu.

**Comment by stdevEclipse** [comment]
I am testing within the skydev container, with the last image you gave me.

**Comment by olemis** [comment]
Oh ! Sorry . My mistake

**Comment by olemis** [comment]
@stdevHan @stdevEclipse I added a new task in this issue , see #1436

**Comment by stdevEclipse** [comment]
All tests involving encryption/decryption that failed with Docker on Windows passed with Docker on Ubuntu. This is probably because Docker on Windows is slower and the service times out.

**Comment by olemis** [comment]
@stdevEclipse remove comments then, make sure they are executed during travis build and push upstream to see the results .

**Comment by stdevEclipse** [comment]
OK. That is pushed already. I need to check travis log, but it worked OK in Ubuntu.
About the other issue, 1- Creating a wallet with a specific seed.
I made the fix in skycoin code and the test passes. I can create an issue and PR for this.

**Comment by stdevHan** [comment]
Tests not done because of wrapper functions were not created:
  - `TestTransactionsFees`
  - `TestSortTransactions`

**Comment by gz-c** [comment]
@stdevHan what is blocking those remaining tests?

**Comment by stdevHan** [comment]
@gz-c  they were functions that had not been exported, but it is already solved

**Comment by olemis** [comment]
@gz-c if we change in #1649 the behavior of `catchApiPanic` then the test suite needs to be updated .

**Comment by olemis** [comment]
@stdevHan `Ripemd160` methods already implemented ?

**Comment by stdevHan** [comment]
@olemis there is no test dedicated to Ripemd160

**Comment by stdevHan** [comment]
@olemis yes, the data type is used, but there is not ripemd160_test.go

**Comment by olemis** [comment]
@stdevHan maybe it's in ./src/cipher/address ?

**Comment by olemis** [comment]
@olemis Ripemd160 tests 

$ grep -nr Test src/ | grep Ripemd
src//cipher/hash_test.go:40:func TestHashRipemd160(t *testing.T) {
src//cipher/hash_test.go:51:func TestRipemd160Set(t *testing.T) {

**Comment by stdevHan** [comment]
if these tests are implemented in libc
https://github.com/simelo/skycoin/blob/stdevHan_t1191_libc_cipher_coin/lib/cgo/tests/check_cipher.hash.c#L25-L78
